ragstack/colbertbase/colbertbase/astra_retriever.py

```diff
@@ -0,0 +1,140 @@
+
+    # Calculate the dot products in a vectorized manner on the GPU
+    sims = torch.matmul(embedding_list, query_vector)
+
+    # Find the maximum similarity (dot product) value
+    max_sim = torch.max(sims)
+
+    # returns a tensor; the item() is the score
+    return max_sim
+
+
+class ColbertAstraRetriever():
+    astra: AstraDB
+    colbertEmbeddings: ColbertTokenEmbeddings
+    verbose: bool
+    is_cuda: bool=False
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        astraDB: AstraDB,
+        colbertEmbeddings: ColbertTokenEmbeddings,
+        verbose: bool=False,
+        **kwargs
+    ):
+        # initialize pydantic base model
+        self.astra = astraDB
+        self.colbertEmbeddings = colbertEmbeddings
+        self.verbose = verbose
+        self.is_cuda = torch.cuda.is_available()
+
+    def retrieve(
+        self,
+        query: str,
+        k: int=10,
+        query_maxlen: int=64,
+        **kwargs
+    ):
+        #
+        # if the query has fewer than a predefined number of of tokens Nq,
+        # colbertEmbeddings will pad it with BERT special [mast] token up to length Nq.
+        #
+        query_encodings = self.colbertEmbeddings.encode_query(query, query_maxlen=query_maxlen)
+
+        count = self.astra.session.execute(self.astra.chunk_counts_stmt).one().count
+        k = min(k, count)
+
+        # the min of query_maxlen is 32
+        top_k = max(math.floor(len(query_encodings) / 2), 16)
+        if self.verbose:
+            print(f"Total number of chunks: {count}, query length {len(query)} embeddings top_k: {top_k}")
+
+        # find the most relevant documents
+        docparts = set()
+        for qv in query_encodings:
+            # per token based retrieval
+            rows = self.astra.session.execute(self.astra.query_colbert_ann_stmt, [list(qv), top_k])
+            docparts.update((row.title, row.part) for row in rows)
+        # score each document
+        scores = {}
+        for title, part in docparts:
+            # find all the found parts so that we can do max similarity search
+            rows = self.astra.session.execute(self.astra.query_colbert_parts_stmt, [title, part])
+            embeddings_for_part = [tensor(row.bert_embedding) for row in rows]
+            # score based on The function returns the highest similarity score
+            #(i.e., the maximum dot product value) between the query vector and any of the embedding vectors in the list.
+            scores[(title, part)] = sum(max_similarity_torch(qv, embeddings_for_part, self.is_cuda) for qv in query_encodings)
+        # load the source chunk for the top k documents
+        docs_by_score = sorted(scores, key=scores.get, reverse=True)[:k]
+        answers = []
+        rank = 1
+        for title, part in docs_by_score:
+            rs = self.astra.session.execute(self.astra.query_part_by_pk_stmt, [title, part])
```

Comment:
i think the retriever approach i took here is faster:
https://github.com/datastax-labs/colbert-wikipedia-data/blob/main/serve.py#L52-L80

Reply:
That's a great comments. execute_async improved about 70% on 50000 chunks search.

Reply:
Good point! I changed to use futures and execute_async. Since Eq is much less than Ed, the improvement is still much better around 20% latency improvement

Reply:
+1 for using async

Reply:
And to add to this, we'll need an asyncio async/await impl at some point.